File: src/386/java/3p.java
import java.util.Arrays;

public class Solution {
    public static int earliestSecondToMarkIndices(int[] nums, int[] changeIndices) {
        int n = nums.length;
        int m = changeIndices.length;
        if (n > m) {
            return -1;
        }

        int[] lastT = new int[n];
        int left = n - 1, right = m - 1;
        int ans = -2;
        while (left <= right) {
            int mid = (left + right) / 2;

            if (check(nums, changeIndices, lastT, mid + 1)) {
                ans = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return ans + 1;
    }

    private static boolean check(int[] nums, int[] changeIndices, int[] lastT, int mx) {
        Arrays.fill(lastT, -1);
        for (int i = 0; i < mx; i++) {
            lastT[changeIndices[i] - 1] = i;
        }

        for (int i : lastT) {
            if (i < 0) { // 有课程没有考试时间
                return false;
            }
        }

        int cnt = 0;
        for (int i = 0; i < mx; i++) {
            int idx = changeIndices[i] - 1;
            if (i == lastT[idx]) {
                if (nums[idx] > cnt) {
                    return false;
                }

                cnt -= nums[idx];
            } else {
                ++cnt;
            }
        }

        return true;
    }
}

File: src/386/cpp/3p.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
 public:
  int earliestSecondToMarkIndices(vector<int> &nums, vector<int> &changeIndices) {
	int n = nums.size(), m = changeIndices.size();
	if (n > m) return -1;

	vector<int> last_t(n);
	auto check = [&](int mx) -> bool {
	  ranges::fill(last_t, -1);

	  for (int t = 0; t < mx; ++t) {
		last_t[changeIndices[t] - 1] = t;
	  }

	  if (ranges::find(last_t, -1) != last_t.end()) {
		return false;
	  }

	  int cnt = 0;
	  for (int i = 0; i < mx; ++i) {
		int idx = changeIndices[i] - 1;

		// 考试
		if (i == last_t[idx]) {
		  if (nums[idx] > cnt) { // 没时间复习
			return false;
		  }

		  cnt -= nums[idx]; // 复习这门课
		} else {
		  ++cnt; // 留着后面用
		}
	  }

	  return true;
	};

	int left = n - 1, right = m - 1;
	int ans = -2;
	while (left <= right) {
	  int mid = (left + right) / 2;

	  // mid + 1 才是 完成所有目标的天数
	  if (check(mid + 1)) {
		ans = mid;
		right = mid - 1;
	  } else {
		left = mid + 1;
	  }
	}

	return ans + 1;
  }
};
